bitcoinspv/submit_test.go

```diff
@@ -0,0 +1,354 @@
+package bitcoinspv
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"testing"
+	"time"
+
+	"github.com/btcsuite/btcd/wire"
+	"github.com/gonative-cc/relayer/bitcoinspv/clients/mocks"
+	sui_errors "github.com/gonative-cc/relayer/bitcoinspv/clients/sui"
+	"github.com/gonative-cc/relayer/bitcoinspv/config"
+	"github.com/rs/zerolog"
+	"github.com/stretchr/testify/assert"
+)
+
+var testSubmitConfig = &config.RelayerConfig{
+	RetrySleepDuration:    1 * time.Millisecond,
+	MaxRetrySleepDuration: 50 * time.Millisecond,
+	HeadersChunkSize:      2,
+	Format:                "auto",
+	Level:                 "debug",
+	NetParams:             "test",
+	BTCCacheSize:          100,
+	BTCConfirmationDepth:  6,
+	ProcessBlockTimeout:   5 * time.Second,
+}
+
+func TestFindFirstNewHeader(t *testing.T) {
+	ctx := context.Background()
+	testBlocks := createTestIndexedBlocks(t, 5, 100) // heights 100, 101, 102, 103, 104
+
+	tests := []struct {
+		name          string
+		mockSetup     func(mockLC *mocks.MockBitcoinSPV)
+		expectedIndex int
+		expectedCalls int
+	}{
+		{
+			name: "all headers exist",
+			mockSetup: func(mockLC *mocks.MockBitcoinSPV) {
+				for _, b := range testBlocks {
+					mockLC.On("ContainsBlock", ctx, b.BlockHash()).Return(true, nil).Once()
+				}
+			},
+			expectedIndex: -1,
+			expectedCalls: 5,
+		},
+		{
+			name: "no headers exist",
+			mockSetup: func(mockLC *mocks.MockBitcoinSPV) {
+				mockLC.On("ContainsBlock", ctx, testBlocks[0].BlockHash()).Return(false, nil).Once()
+			},
+			expectedIndex: 0,
+			expectedCalls: 1,
+		},
+		{
+			name: "some headers exist",
+			mockSetup: func(mockLC *mocks.MockBitcoinSPV) {
+				mockLC.On("ContainsBlock", ctx, testBlocks[0].BlockHash()).Return(true, nil).Once()
+				mockLC.On("ContainsBlock", ctx, testBlocks[1].BlockHash()).Return(true, nil).Once()
+				mockLC.On("ContainsBlock", ctx, testBlocks[2].BlockHash()).Return(false, nil).Once()
+			},
+			expectedIndex: 2,
+			expectedCalls: 3,
+		},
+		{
+			name: "ContainsBlock retryable error then success (finds new)",
+			mockSetup: func(mockLC *mocks.MockBitcoinSPV) {
+				retryableErr := errors.New("network timeout")
+				mockLC.On("ContainsBlock", ctx, testBlocks[0].BlockHash()).Return(true, nil).Once()
+				mockLC.On("ContainsBlock", ctx, testBlocks[1].BlockHash()).Return(false, retryableErr).Once() // retry after error
+				mockLC.On("ContainsBlock", ctx, testBlocks[1].BlockHash()).Return(true, nil).Once()           // success on retry
+				mockLC.On("ContainsBlock", ctx, testBlocks[2].BlockHash()).Return(false, nil).Once()
+			},
+			expectedIndex: 2,
+			expectedCalls: 4,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			mockLC := mocks.NewMockBitcoinSPV(t)
+			tt.mockSetup(mockLC)
+			r := &Relayer{
+				lcClient: mockLC,
+				logger:   zerolog.Nop(),
+				Config:   testSubmitConfig,
+			}
+			idx, err := r.findFirstNewHeader(ctx, testBlocks)
```

Comment:
The findFirstNewHeader name doesn't make sense. findFirstNewHeader means the first header but we don't know the context. Maybe we can rename to find firstHeaderNonSync or  FindFirstNewestBlock.

Reply:
Updated